### ISSUE TITLE
Keychain error

### ISSUE THREAD

**Issue by startupthekid** [forcedotcom/ServiceSDK-iOS#11]
The release notes say (under known issues):

"If your app is launched in the background while the device is locked (for example, in response to an actionable push notification or a Core Location geofence event), the application may crash as a result of accessing a locked keychain entry. Workaround: Please contact Salesforce support for a workaround"

Is there any progress on this at all? We contacted support months ago and this crash is responsible for the large majority of crashes we see in the field (it's been our top crash forever). Is there any update to this? Can you please fix it?

**Comment by quidcomobile** [comment]
Hi there,

We have exactly the same issue. We're even having a surge in number of crashes starting a few hours ago after we released a new version of our app 2 days ago (planned over 7 days release, it's 2% now but we have hundreds of crashes).

**Comment by gordienoye** [comment]
We have a workaround for this issue and we are working on the permanent fix. I have posted a dropbox link which contains a zipped file. All you should need to do is add the MobileSDKLaunchFix.m file to your application and make sure it's compiled into your application, and the fix should be applied immediately.

[dropbox.com - MobileSDKLaunchFix.zip](https://www.dropbox.com/s/rmggqdeeafg6g49/MobileSDKLaunchFix.zip?dl=0)

The next release of the SDKs in February will address the underlying issue and this temporary fix should be removed.

**Comment by quidcomobile** [comment]
Thanks @gordienoye,

In fact, I found the workaround on Google+ Community about one hour ago
https://plus.google.com/communities/104189600084591024148/s/keychain

So, we're going use it, but I I'd like to know if setting sKeychainAccessExceptionsAreFatal to false would also fix this.

We are very tempted to do both.

**Comment by startupthekid** [comment]
Uh I'm sorry, but I'm not about to download a zip from some unknown developer and then proceed to put it into my company's project.  Neither should anyone for that matter. @gordienoye do you work at salesforce?

**Comment by quidcomobile** [comment]
@startupthekid I don't know if @gordienoye works at Salesforce, but **Karthik Kannappan** (is he working at Salesforce?), published on Google+ the same workaround, including this note:

> We have a workaround for this issue and we are working on the permanent fix. I have posted a dropbox link which contains a zipped file. All you should need to do is add the MobileSDKLaunchFix.m file to your application and make sure it's compiled into your application, and the fix should be applied immediately.

You can read the full post at:
https://plus.google.com/communities/104189600084591024148/s/keychain

In general, I fully agree with you - I should wait for a new SDK rather than putting this in the project, but I haven't the luxury atm, and the workaround seems pretty safe to use. So we are going to use it - or rather, adapt it to our own needs.

We'll remove this when the new SDK is available though.

**Comment by gordienoye** [comment]
Hi @quidcomobile and @startupthekid 

I do work at Salesforce on the Snap-ins SDKs. I can appreciate not wanting to download a zip file so I will include the .m file here. We are fixing the underlying issue that caused the problem for the February release and the MobileSDKLaunchFix.m file should be removed with the next version.

```
//
//  MobileSDKLaunchFix.m
//  ServiceSDKTestApp
//
//  Created by Michael Nachbaur on 1/11/18.
//  Copyright © 2018 Salesforce.com. All rights reserved.
//

#import <objc/runtime.h>
#import <UIKit/UIKit.h>
@import SalesforceSDKCore.SFKeychainItemWrapper;

static id kWillEnterForegroundHandle = nil;

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wundeclared-selector"
__attribute__((constructor))
static void swizzle_sdk_initialization() {
    // Retrieve the current implementation of the Snap-ins initialization method which normally
    // is invoked by UIApplicationDidFinishLaunchingNotification.
    Class SCSWindowManagerClass = NSClassFromString(@"SCSWindowManager");
    SEL applicationFinishedLaunchingSelector = @selector(applicationFinishedLaunching:);
    Method method = class_getInstanceMethod(SCSWindowManagerClass, applicationFinishedLaunchingSelector);

    // Replace the default implementation of that method with one that saves the reference to the SCSWindowManager instance
    // that was to perform the initialization.
    __block id _windowManager = nil;
    IMP originalImplementation = method_setImplementation(method, imp_implementationWithBlock(^(__unsafe_unretained id _self, __unsafe_unretained NSNotification *notification) {
        _windowManager = _self;
    }));

    // Register a notification observer for UIApplicationWillEnterForegroundNotification to invoke the Snap-ins initialization
    // once the application is in the foreground.  This ensures that the device is unlocked, and the keychain is available
    // to the application.
    kWillEnterForegroundHandle = [[NSNotificationCenter defaultCenter] addObserverForName:UIApplicationWillEnterForegroundNotification object:nil queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification * _Nonnull note) {
        // Remove the observer so this method can only be invoked once during the lifetime of the application.
        // This ensures the initialization work is only performed once.
        [[NSNotificationCenter defaultCenter] removeObserver:kWillEnterForegroundHandle];

        @try {
            // Change the MobileSDK's default keychain access policy to a more permissive mode,
            // so the keychain can be accessible in the background in the future.
            [SFKeychainItemWrapper setAccessibleAttribute:kSecAttrAccessibleAfterFirstUnlock];

            // Invoke the original implementation of the initialization method, allowing us to
            // defer the initialization once the application is in the foreground.
            ((void(*)(id, SEL, id))originalImplementation)(_windowManager, applicationFinishedLaunchingSelector, note);
        } @catch (NSException *e) {
            NSLog(@"Tried to initialize the MobileSDK runtime, but an exception occurred: %@", e);
        }
    }];
}
#pragma clang diagnostic pop
```

**Comment by startupthekid** [comment]
Is the release coming out any time soon? There's not many days left for a February release...

**Comment by sfyz** [comment]
We’re still targeting a late February release. I’ll follow up next week with an update.

**Comment by startupthekid** [comment]
Just saw there's a new release! 🎉 🎉 does that fix this issue?

**Comment by sfyz** [comment]
@startupthekid Yes, I believe this fix should address your issue. Keep us posted if you run into further problems.
